src/main/java/org/apache/xml/security/encryption/keys/content/derivedKey/HKDF.java

```diff
@@ -0,0 +1,182 @@
+     * @param prk                  pseudo-random key derived from the shared secret
+     * @param info                 used to derive the key
+     * @param keyLength            key length in bytes of the derived key
+     * @return the output keying material (OKM) size of keyLength octets
+     * @throws XMLSecurityException if the jceHmacAlgorithmName is not supported
+     */
+    public byte[] expandKey(String jceHmacAlgorithmName, byte[] prk, byte[] info, long keyLength) throws XMLSecurityException {
+        // prepare for expanding the key
+        Mac hMac = initHMac(jceHmacAlgorithmName, prk, false);
+        int iMacLength = hMac.getMacLength();
+
+        int toGenerateSize = (int) keyLength;
+        ByteBuffer result = ByteBuffer.allocate(toGenerateSize);
+        byte[] prevResult = new byte[0];
+        short counter = 1;
+        while (toGenerateSize > 0) {
+            hMac.reset();
+            hMac.update(prevResult);
+            if (info != null && info.length > 0) {
+                hMac.update(info);
+            }
+            hMac.update((byte) counter);
+            prevResult = hMac.doFinal();
+            result.put(prevResult, 0, Math.min(toGenerateSize, iMacLength));
+            // get ready for next iteration
+            toGenerateSize -= iMacLength;
+            counter++;
+        }
+        return result.array();
+    }
+
+    /**
+     * Method initializes a Message Authentication Code (MAC) object using the
+     * init secret/salt or an empty byte array if initSecret parameter is null or empty.
+     *
+     * @param jceAlgorithmName the java JCE HMAC algorithm name to use to init Mac
+     * @param initSecret       the secret/salt to initialize the hmac
+     * @param initPRK          if true, the salt is initialized with a string of zero octets
+     *                         as long as the hash function output see [RFC5869] Section 2.2
+     * @return Initialized Mac object
+     * @throws XMLSecurityException if the hmac algorithm is not supported or if it
+     *  fails to initialize
+     */
+    private Mac initHMac(String jceAlgorithmName, byte[] initSecret, boolean initPRK) throws XMLSecurityException {
+        Mac mac;
+        try {
+            LOG.log(DEBUG, "Init Mac with hash algorithm: [{}]", jceAlgorithmName);
+            mac = Mac.getInstance(jceAlgorithmName);
+        } catch (NoSuchAlgorithmException e) {
+            throw new XMLSecurityException(e, "KeyDerivation.NotSupportedParameter", new Object[]{jceAlgorithmName});
+        }
+
+        if (initPRK && (initSecret == null || initSecret.length == 0)) {
+            //  If "initSecret"/salt is not provided, a string of zero octets as long as the hash function output is used
+            LOG.log(DEBUG, "Init Mac with hmac algorithm [{}] and empty salt!", jceAlgorithmName);
+            initSecret = new byte[mac.getMacLength()];
+        }
+        SecretKeySpec secretKey = new SecretKeySpec(initSecret, jceAlgorithmName);
```

Comment:
@coheigea, @seanjmullan, I would need your advice on this check failure. The  method  "XMLCipherUtil.getJCEMacHashForUri" is just mapping URI names to JCE names. Should I just log warning for HmacSHA1, HmacSHA224, HmacSHA384, HMACRIPEMD160 or should the method throw and error since algorithms are weak or should we just ignore the CodeQL?

Reply:
A few comments/questions:

- Why are all of these considered weak/insecure? Ok, I understand HmacSHA1 could be considered weak, but why HmacSHA224, HmacSHA384?
- There is already a secureValidation mode implemented in Santuario which restricts weak crypto algorithms; however HmacSHA1 is not disabled yet. We could consider disabling it, but we also need to assess the compatibility risk of doing that, which may be too high. 
- There may be some more places we should be adding additional secureValidation checks in the HKDF code. Can you explore this further?